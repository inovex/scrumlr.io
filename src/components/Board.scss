@import "common";

.board {
  --n: 1;

  display:grid;
  grid-auto-flow: column;
  grid-auto-columns: calc(100%/min(var(--real-columns), var(--n)));
  overflow-x:auto;
  scroll-snap-type: x mandatory;

  height: 100vh;
}

@for $i from 1 through 12 {
  @media screen and (min-width: $i*$min-column-width) {
    .board {
      --n: #{$i};
    }
  }
}