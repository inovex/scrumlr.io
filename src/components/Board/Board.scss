@import "constants/style";
@import "components/Column/Column";

.board {
  --n: 1;

  display: grid;
  grid-auto-flow: column;
  grid-template-columns:
    minmax($board__side-panel-width, 1fr)
    repeat(var(--board__columns), Min(calc((100% - #{$board__side-panel-width * 2}) / min(var(--board__columns), var(--n))), #{$column__max-width}))
    minmax($board__side-panel-width, 1fr);

  scroll-snap-type: x mandatory;
  scroll-snap-stop: always;
  scroll-padding: $board__side-panel-width;
  overflow-x: auto;

  width: 100vw;
  height: 100vh;

  -ms-overflow-style: none;
  scrollbar-width: none;
}

.board::-webkit-scrollbar {
  display: none;
}

.board > * {
  scroll-snap-align: start;
}

.board__spacer-left {
  @include inset-border($top: true, $left: true, $bottom: true);
}
.board__spacer-right {
  @include inset-border($top: true, $right: true, $bottom: true);
}

.board__spacer-left:nth-child(odd),
.board__spacer-right:nth-child(odd) {
  background-color: #f9fafb;
}

.board__spacer-left:nth-child(even),
.board__spacer-right:nth-child(even) {
  background-color: #f6f5fb;
}

@media (prefers-color-scheme: dark) {
  .board__spacer-left:nth-child(odd),
  .board__spacer-right:nth-child(odd) {
    background-color: #242c3d;
  }

  .board__spacer-left:nth-child(even),
  .board__spacer-right:nth-child(even) {
    background-color: #272f41;
  }
}

$board__navigation-button-dimension: 74px;
$board__navigation-button-icon-dimension: 24px;

.board__navigation {
  position: fixed;
  top: calc(50% - #{$board__navigation-button-dimension/2});

  display: flex;
  flex-direction: column;
  justify-content: center;

  width: $board__navigation-button-dimension/2;
  height: $board__navigation-button-dimension;

  background-color: var(--accent-color);
  opacity: 50%;
  transition: opacity 0.15s ease-in-out;

  border: none;

  margin: 0;
  padding: 0;

  z-index: $board__navigation-button-z-index;
}

.board__navigation:focus {
  outline: none;
}

.board__navigation:hover {
  opacity: 100%;
}

.board__navigation-prev {
  left: 0;

  // border-radius enables half circle
  border-radius: 0 74px 74px 0;
}
.board__navigation-next {
  right: 0;

  // border-radius enables half circle
  border-radius: 74px 0 0 74px;
}

.board__navigation-arrow {
  height: $board__navigation-button-icon-dimension;
  width: $board__navigation-button-icon-dimension;
  fill: $color-white;
}

.board__navigation-arrow-prev {
  align-self: flex-start;

  // visually center svg by margin
  margin-left: 3px;
}
.board__navigation-arrow-next {
  align-self: flex-end;

  // visually center svg by margin
  margin-right: 3px;
}

@media screen and (max-width: $column__min-width+2*$board__side-panel-width) {
  .board {
    grid-template-columns: $column__border-width repeat(var(--board__columns), calc(100% - #{$column__border-width * 2})) $column__border-width;
    scroll-padding: $column__border-width;
  }
}

@for $i from 1 through 16 {
  @media screen and (min-width: $i*($column__min-width+2*$board__side-panel-width)) {
    .board {
      --n: #{$i};
    }
  }
}
