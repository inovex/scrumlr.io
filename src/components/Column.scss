@import "../constants/style";

$border__size: 8px;

$border__left: inset #{2*$border__size} 0 0 #{-$border__size} var(--accent-color);
$border__right: inset #{-2*$border__size} 0 0 #{-$border__size} var(--accent-color);
$border__top: inset 0 #{2*$border__size} 0 #{-$border__size} var(--accent-color);
$border__bottom: inset 0 #{-2*$border__size} 0 #{-$border__size} var(--accent-color);

.column {
  display: flex;
  justify-content: center;

  width: 100%;
  height: 100%;
  scroll-snap-align: start;

  box-shadow: $border__top, $border__bottom;
}

.column:only-of-type {
  box-shadow: $border__top, $border__bottom, $border__left, $border__right;
}
.column:not(:only-of-type):first-of-type {
  box-shadow: $border__top, $border__bottom, $border__left;
}
.column:not(:only-of-type):last-of-type {
  box-shadow: $border__top, $border__bottom, $border__right;
}

.column__content {
  flex: 1 0 $min-column-width;
  min-width: 320px;
  max-width: 1440px;
}