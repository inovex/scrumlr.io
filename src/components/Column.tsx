import React, { useContext, useState } from 'react';
import TextField from '@material-ui/core/TextField';
import Button from '@material-ui/core/Button';
import { BoardContext } from '../routing/pages/BoardContext';
import CardComponent from './Card';
import { addCard } from '../domain/board';
import { WithId } from '../util/withId';
import { Card } from '../types/state';

export interface ColumnContextType {
  columnId?: string;
}

export const ColumnContext = React.createContext<ColumnContextType>({
  columnId: undefined,
});

export interface ColumnProps {
  id: string;
  name: string;
  visible: boolean;
  cards: WithId<Card>[];
}

export interface ColumnState {
  text?: string;
}

export const Column: React.FC<ColumnProps> = ({ id, name, cards }) => {
  const [state, setState] = useState<ColumnState>({});
  const { boardId, isAdmin } = useContext(BoardContext);

  const onAddCard = () => {
    addCard(boardId!, id, state.text).then(() => {
      setState({ ...state, text: undefined });
    });
  };

  return (
    <ColumnContext.Provider value={{ columnId: id }}>
      <div>
        <p>{name}</p>

        {isAdmin && (
          <p>
            <Button>Toggle Visibility</Button>
          </p>
        )}

        <TextField label="Add card" value={state.text || ''} onChange={(event) => setState({ ...state, text: event.target.value })} />
        <Button onClick={onAddCard}>Add card</Button>
        <ul>
          {cards.map((card) => (
            <li key={card.id}>
              <CardComponent {...card} />
            </li>
          ))}
        </ul>
      </div>
    </ColumnContext.Provider>
  );
};

export default Column;
