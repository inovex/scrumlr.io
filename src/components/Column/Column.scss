@import "constants/style";

.column {
  min-width: $column__min-width;
  width: 100%;
  max-width: $column__max-width;
  @include inset-border($top: true, $bottom: true);
  padding-top: $header__height;
  height: 100vh;
}
.column:nth-child(even) {
  background-color: #f9fafb;
}
.column:nth-child(odd) {
  background-color: #f6f5fb;
}
@media (prefers-color-scheme: dark) {
  .column:nth-child(even) {
    background-color: #242c3d;
  }
  .column:nth-child(odd) {
    background-color: #272f41;
  }
}

.column__content {
  max-height: 100%;
  padding: $padding--large 0 $column__border-width 0;
  display: flex;
  flex-direction: column;
}

.column__header {
  padding: 0 $padding--large;
}

.column__header-title {
  display: flex;
}

.column__header-text {
  font-family: Raleway;
  font-size: 24px;
  color: $color-black;
  border-bottom: solid 3px var(--accent-color);
  margin-right: 10px;
  font-weight: bold;
  letter-spacing: $letter-spacing--large;
  line-height: $line-height--large;
}
@media (prefers-color-scheme: dark) {
  .column__header-text {
    color: $color-white;
  }
}

.column__header-card-number {
  font-family: Raleway;
  font-size: 20px;
  color: $color-middle-gray;
  font-weight: 600;
  letter-spacing: $letter-spacing--large;
  line-height: $line-height--large;
}

.column__notes-wrapper {
  overflow: auto;
}

.column__notes-wrapper::-webkit-scrollbar {
  width: 12px;
}

.column__notes-wrapper::-webkit-scrollbar-track {
  -webkit-box-shadow: inset 0 0 6px $color-light-gray;
  border-radius: 10px;
}

.column__notes-wrapper::-webkit-scrollbar-thumb {
  border-radius: 10px;
  background-color: $color-middle-gray;
}

@media (prefers-color-scheme: dark) {
  .column__notes-wrapper::-webkit-scrollbar-track {
    -webkit-box-shadow: inset 0 0 6px $color-dark-mode;
    background-color: $color-dark-mode-note;
  }
  .column__notes-wrapper::-webkit-scrollbar-thumb {
    border-radius: 10px;
    background-color: $color-darkest-gray;
  }
}

.column__note-list {
  margin: 0;
  padding: 0 $padding--large $padding--default $padding--large;
  display: grid;
  grid-auto-flow: row;
  grid-row-gap: 20px;
  grid-column-gap: 24px;
}
