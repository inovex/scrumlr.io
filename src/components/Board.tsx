import React from 'react';
import './Board.scss';
import {getColorClassName} from "../constants/colors";

const Board = ({ children }: any) => {
    const colors = React.Children.map(children, (child) => child.props.color);

    return (
        <>
            <style>
                {`.board { --board__columns: ${React.Children.count(children)} }`}
            </style>
            <div className="board">
                <div className={`board__spacer-left ${getColorClassName(colors[0])}`} />
                    {children}
                <div className={`board__spacer-right ${getColorClassName(colors[colors.length - 1])}`} />
            </div>
        </>
    )
};

export default Board;