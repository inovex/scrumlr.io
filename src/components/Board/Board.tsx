import React, {useEffect, useRef, useState} from "react";
import {getColorClassName} from "constants/colors";
import {ColumnProps} from "components/Column/Column";
import {ReactComponent as RightArrowIcon} from "assets/icon-arrow-next.svg";
import {ReactComponent as LeftArrowIcon} from "assets/icon-arrow-previous.svg";
import MenuBars from "components/MenuBars/MenuBars";
import BoardHeader from "components/BoardHeader/BoardHeader";
import "./Board.scss";

export interface BoardProps {
  children: React.ReactElement<ColumnProps> | React.ReactElement<ColumnProps>[];
  name: string;
  boardstatus: string;
}

export interface BoardState {
  firstVisibleColumnIndex: number;
  lastVisibleColumnIndex: number;
}

const Board = ({children, name, boardstatus}: BoardProps) => {
  const [state, setState] = useState<BoardState>({firstVisibleColumnIndex: 0, lastVisibleColumnIndex: React.Children.count(children)});
  const boardRef = useRef<HTMLDivElement>(null);
  const columnVisibilityStatesRef = useRef<boolean[]>([]);
  const intersectionObserverRef = useRef<IntersectionObserver | null>(null);

  useEffect(() => {
    const board = boardRef.current;

    // disconnect the previous observer, if there is one
    if (intersectionObserverRef.current !== null) {
      intersectionObserverRef.current.disconnect();
    }

    if (board) {
      // initialize column visibility states
      columnVisibilityStatesRef.current = new Array(React.Children.count(children));
      const columnVisibilityStates = columnVisibilityStatesRef.current;
      columnVisibilityStates.fill(false);

      // initialize intersection observer
      const observerOptions = {
        root: board,
        rootMargin: "0px",
        threshold: 1.0,
      };
      const observerCallback: IntersectionObserverCallback = (entries) => {
        entries.forEach((entry) => {
          const index = Array.prototype.indexOf.call(board.children, entry.target) - 1;
          columnVisibilityStates[index] = entry.isIntersecting;
        });
        setState({
          firstVisibleColumnIndex: columnVisibilityStates.findIndex((value) => value),
          lastVisibleColumnIndex: columnVisibilityStates.lastIndexOf(true),
        });
      };
      const observer = new IntersectionObserver(observerCallback, observerOptions);

      // observe children
      const domChildren = board.children;
      for (let i = 1; i < domChildren.length - 1; i += 1) {
        observer.observe(domChildren[i]);
      }

      // return callback handler that will disconnect the observer on unmount
      return () => {
        observer.disconnect();
      };
    }
    return undefined;
  }, [children]);

  const columnsCount = React.Children.count(children);
  if (!children || columnsCount === 0) {
    return <div className="board--empty">Empty board</div>;
  }

  const {firstVisibleColumnIndex, lastVisibleColumnIndex} = state;
  const columnColors = React.Children.map(children, (child) => child.props.color);

  const showNextButton = lastVisibleColumnIndex < columnsCount - 1;
  const showPreviousButton = firstVisibleColumnIndex > 0;

  const previousColumnIndex = firstVisibleColumnIndex > 0 ? firstVisibleColumnIndex - 1 : columnsCount - 1;
  const nextColumnIndex = lastVisibleColumnIndex === columnsCount - 1 ? 0 : firstVisibleColumnIndex + 1;

  const handlePreviousClick = () => {
    boardRef.current!.children[previousColumnIndex + 1].scrollIntoView({inline: "start", behavior: "smooth"});
  };

  const handleNextClick = () => {
    boardRef.current!.children[nextColumnIndex + 1].scrollIntoView({inline: "start", behavior: "smooth"});
  };

  return (
    <>
      <style>{`.board { --board__columns: ${columnsCount} }`}</style>

      <BoardHeader name={name} boardstatus={boardstatus} />
      <MenuBars />

      {showPreviousButton && (
        <button className={`board__navigation board__navigation-prev ${getColorClassName(columnColors[previousColumnIndex])}`} onClick={handlePreviousClick} aria-hidden>
          <LeftArrowIcon className="board__navigation-arrow board__navigation-arrow-prev" />
        </button>
      )}

      <main className="board" ref={boardRef}>
        <div className={`board__spacer-left ${getColorClassName(columnColors[0])}`} />
        {children}
        <div className={`board__spacer-right ${getColorClassName(columnColors[columnColors.length - 1])}`} />
      </main>

      {showNextButton && (
        <button
          className={`board__navigation board__navigation-next ${getColorClassName(columnColors[(lastVisibleColumnIndex + 1) % columnColors.length])}`}
          onClick={handleNextClick}
          aria-hidden
        >
          <RightArrowIcon className="board__navigation-arrow board__navigation-arrow-next" />
        </button>
      )}
    </>
  );
};

export default Board;
