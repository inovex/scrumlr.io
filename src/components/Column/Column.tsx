import * as classNames from 'classnames';
import * as React from 'react';

import './Column.scss';
import { Card as CardModel } from '../../types';
import {
  ColumnConfiguration,
  IndexedPhaseConfiguration
} from '../../constants/Retrospective';
import { mapStateToProps } from './Column.container';
import FocusedCardComponent from './FocusedCardComponent';
import StackComponent from './StackComponent';
import ColumnHeader from './ColumnHeader';
import * as ReactCSSTransitionGroup from 'react-transition-group/CSSTransitionGroup';
import { Icon } from '../Icon';
import { connect } from 'react-redux';
import { ColumnOverview } from '../ColumnOverview';

export interface OwnColumnProps {
  boardUrl: string;
  isAdmin: boolean;

  /** The unique column type. */
  column: ColumnConfiguration;
  phase: IndexedPhaseConfiguration;

  isActive: boolean;

  hasNextColumn?: boolean;
  hasPreviousColumn?: boolean;
  onGoToNextColumn?: () => void;
  onGoToPrevColumn?: () => void;
  isCompactView: boolean;

  className?: string;
}

export interface StateColumnProps {
  theme: string;
  cards: CardModel[];
  focused?: CardModel;
  isHidden: boolean;
  isExtended: boolean;
}

export type ColumnProps = OwnColumnProps & StateColumnProps;

export interface ColumnState {
  showOverview: boolean;
}

export class Column extends React.Component<ColumnProps, ColumnState> {
  state: ColumnState = {
    showOverview: false
  };

  toggleOverview = () => {
    this.setState({ showOverview: !this.state.showOverview });
  };

  render() {
    const {
      isAdmin,
      isActive,
      isHidden,
      isExtended,
      className,
      focused,
      column,
      phase,
      theme,
      boardUrl,
      cards,
      isCompactView
    } = this.props;

    const cardsCount = cards.filter(c => !c.parent).length;

    return (
      <>
        {this.state.showOverview && (
          <ColumnOverview
            boardUrl={boardUrl}
            column={column.name}
            cardsCount={cardsCount}
            cards={cards}
            isVotingEnabled={column.voting.enabled}
            isVoteSummaryShown={column.voting.displayed}
            toggleOverview={this.toggleOverview}
          />
        )}
        <div
          className={classNames('column', `column--theme-${theme}`, className, {
            ['column--inactive']:
              !isActive && (focused ? focused.type !== column.id : true),
            ['column--hidden']: isHidden
          })}
          style={{ flex: isExtended ? phase.columns.length - 1 : 1 }}
        >
          {!focused && this.props.hasPreviousColumn && (
            <button
              className={classNames(
                'column__navigation',
                'column__navigation-left'
              )}
              type="button"
              onClick={this.props.onGoToPrevColumn}
              disabled={!isActive}
            >
              <Icon name="chevron-left" width={32} height={32} />
            </button>
          )}

          {!focused && this.props.hasNextColumn && (
            <button
              className={classNames(
                'column__navigation',
                'column__navigation-right'
              )}
              type="button"
              onClick={this.props.onGoToNextColumn}
              disabled={!isActive}
            >
              <Icon name="chevron-right" width={32} height={32} />
            </button>
          )}

          <ColumnHeader
            title={column.name}
            count={cardsCount}
            onToggleOverview={this.toggleOverview}
          />

          <ReactCSSTransitionGroup
            transitionName="column__content-animation"
            transitionEnterTimeout={600}
            transitionEnter={true}
            transitionLeave={false}
            className="column__content"
          >
            {focused && column.focus.align === 'left' && (
              <FocusedCardComponent
                isAdmin={isAdmin}
                key={focused.id}
                boardUrl={boardUrl}
                focused={focused}
                className="component--large"
                showVotes={column.voting.displayed}
              />
            )}

            {!isExtended && (
              <StackComponent
                boardUrl={boardUrl}
                cards={cards}
                isVotingAllowed={column.voting.enabled}
                isVoteSummaryShown={column.voting.displayed}
                column={column}
                isActive={isActive || !isCompactView}
                className={classNames('column__stack-component', {
                  ['column__stack-component--hidden']: Boolean(focused)
                })}
              />
            )}

            {focused && column.focus.align === 'right' && (
              <FocusedCardComponent
                isAdmin={isAdmin}
                key={focused.id}
                boardUrl={boardUrl}
                focused={focused}
                className="component--large"
                showVotes={column.voting.displayed}
              />
            )}
          </ReactCSSTransitionGroup>
        </div>
      </>
    );
  }
}

export default connect<StateColumnProps, null, OwnColumnProps>(mapStateToProps)(
  Column
);
