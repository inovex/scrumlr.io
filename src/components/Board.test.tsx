import React from 'react'
import {render} from "@testing-library/react";
import Board from "./Board";
import Column from "./Column";

test('show empty board', () => {
    const { container } = render(
        <Board>
        </Board>
    );

    expect(container.firstChild).toHaveClass("board--empty");
});

test('correct number of columns is set in inner styles', () => {
    const { container } = render(
        <Board>
            <Column color="violet" />
            <Column color="pink" />
            <Column color="blue" />
            <Column color="purple" />
        </Board>
    );

    expect(container.querySelector('style')).toHaveTextContent('.board { --board__columns: 4 }');
});

describe('side-panels', () => {
    test('left side-panel is present', () => {
        const { container } = render(
            <Board>
                <Column color="blue" />
                <Column color="pink" />
            </Board>
        );
        expect(container.querySelector(".board").firstChild).toHaveClass("board__spacer-left");
    });

    test('right side-panel is present', () => {
        const { container } = render(
            <Board>
                <Column color="blue" />
                <Column color="pink" />
            </Board>
        );
        expect(container.querySelector(".board").lastChild).toHaveClass("board__spacer-right");
    });

    test('left side-panel has correct accent color', () => {
        const { container } = render(
            <Board>
                <Column color="blue" />
                <Column color="pink" />
            </Board>
        );
        expect(container.querySelector(".board").firstChild).toHaveClass("accent-color__blue");
    });

    test('right side-panel has correct accent color', () => {
        const { container } = render(
            <Board>
                <Column color="blue" />
                <Column color="pink" />
            </Board>
        );
        expect(container.querySelector(".board").lastChild).toHaveClass("accent-color__pink");
    });

    test('side-panels have correct accent color with single column', () => {
        const { container } = render(
            <Board>
                <Column color="violet" />
            </Board>
        );

        const board = container.querySelector(".board");
        expect(board.firstChild).toHaveClass("accent-color__violet");
        expect(board.lastChild).toHaveClass("accent-color__violet");
    });
});

