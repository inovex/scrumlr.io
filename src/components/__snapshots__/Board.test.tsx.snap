// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshot of board 1`] = `
<div>
  <style>
    .board { --board__columns: 4 }
  </style>
  <div
    class="board"
  >
    <div
      class="board__spacer-left accent-color__purple"
    />
    <div
      class="column accent-color__purple"
    >
      <div
        class="column__content"
      />
    </div>
    <div
      class="column accent-color__blue"
    >
      <div
        class="column__content"
      />
    </div>
    <div
      class="column accent-color__violet"
    >
      <div
        class="column__content"
      />
    </div>
    <div
      class="column accent-color__pink"
    >
      <div
        class="column__content"
      />
    </div>
    <div
      class="board__spacer-right accent-color__pink"
    />
  </div>
</div>
`;
